src/GitLabHealth-Model-Importer/GitlabModelImporter.class.st

```diff
@@ -1,1569 +1,1642 @@
+	('Import diff of commit: ' , aCommit short_id printString) recordInfo.
+
+	result := self repoApi commits
+		          diffOf: aCommit id
+		          inProject: aCommit repository project id
+		          uniDiff: true.
+
+	(self isServerError: result) ifTrue: [ ^ {  } ].
+	diffsResult := self newParseDiffResult: result.
+
+	aCommit diffs addAll: diffsResult unless: self blockForDiffEquality.
+	
+	"changes are added into the model during the import"
+	aCommit diffs do: [ :diff | self importDiffRangesForDiff: diff ].
+
+	^ aCommit diffs
+]
+
+{ #category : #'import - merge request' }
+GitlabModelImporter >> importDiffOfMergeRequest: aMergeRequest [
+
+	| result diffsResult |
+	aMergeRequest diffs ifNotEmpty: [
+		'Diff of already importer: '
+		, aMergeRequest iid printString recordInfo.
+		^ aMergeRequest diffs ].
+	('Import diff commits of MR ' , aMergeRequest iid printString)
+		recordInfo.
+	result := self repoApi mergeRequests diffsOf: aMergeRequest iid inProject: aMergeRequest project_id.
+
+	diffsResult := self newParseDiffResult: result.
+
+
+	aMergeRequest diffs
+		addAll: diffsResult
+		unless: self blockForDiffEquality.
+	self glhModel
+		addAll: aMergeRequest diffs
+		unless: self blockForDiffEquality.
+
+	aMergeRequest diffs do: [ :diff | self importDiffRangesForDiff: diff ].
+
+	^ aMergeRequest diffs
+]
+
+{ #category : #'import - repositories' }
+GitlabModelImporter >> importDirectoryFiles: aDirectoryFile OfBranch: aBranch [
+
+	| result files apiFiles params |
+	params := { 
+		#ref -> aBranch name.
+		#path -> (aDirectoryFile path , '/')
+	} asDictionary.
+	result := self repoApi repositories repositoryTreeOfProject: aBranch repository project id withParams: params.
+		         " treeOfRepository: aBranch repository project id
+		          ofBranch: aBranch name
+		          andPath: aDirectoryFile path , '/'."
+	apiFiles := (result collect: [ :treeJson | self parseFileTreeResult: treeJson ]) flattened.
+	files := apiFiles collect: [ :apiFile |
+		         self convertApiFileAsFile: apiFile ].
+	
+	files do: [ :file |
+		self glhModel add: file.
+		aDirectoryFile addFile: file ].
+	
+	files
+		select: [ :file | file isKindOf: GLHFileDirectory ]
+		thenCollect: [ :file |
+		self importDirectoryFiles: file OfBranch: aBranch ]
+]
+
+{ #category : #'import - repositories' }
+GitlabModelImporter >> importFilesOfBranch: aBranch [
+
+	| result files apiFiles params |
+	params := { 
+		#ref -> aBranch name.
+	} asDictionary.
+	
+	result := self repoApi repositories repositoryTreeOfProject: aBranch repository project id withParams: params.
+	
+		          "treeOfRepository: aBranch repository project id
+		          ofBranch: aBranch name
+		          andPath: nil."
+	apiFiles := (result collect: [ :filesJson | self parseFileTreeResult: filesJson  ]) flattened.
+	files := apiFiles collect: [ :apiFile | 
+		         self convertApiFileAsFile: apiFile ].
+	files do: [ :file | 
+		self glhModel add: file.
+		aBranch addFile: file ].

```

Comment:
replace with:  
```smalltalk
files := self glhModel addAll: files unless: [:f1 :f2 | f1 branch = f2 branch and: [f1 path = f2 path]].
files := aBranch addAll: files unless: [:f1 :f2 | f1 branch = f2 branch and: [f1 path = f2 path]].
```
this ensure that a file instance is only added once inside the model